File: lib/rust/prelude/runner.rs
use std::{path::Path, time::Duration};

use crate::prelude::{
    solution::Solution,
    types::{ProblemInput, ProblemResult},
};

use clap::{Args, Parser, Subcommand};

pub struct AocRunner {
    pub year: u16,
    pub solutions: Vec<Box<dyn Solution>>,
}

#[derive(Parser)]
#[command(name = "Advent of Code")]
#[command(about = "Run Advent of Code solutions", long_about = None)]
pub struct Cli {
    #[command(subcommand)]
    pub command: Commands,
}

#[derive(Subcommand)]
pub enum Commands {
    /// Run all implemented days (ignores all options)
    Main,
    /// Run a specific day
    Day(DayArgs),
}

#[derive(Args)]
pub struct DayArgs {
    /// Day number (1-25)
    pub day: u8,

    /// Use sample input instead of real input
    #[arg(short, long)]
    pub sample: bool,

    /// Run only part 1 or part 2 (required unless --all)
    #[arg(short, long, value_parser = clap::value_parser!(u8).range(1..=2))]
    pub part: Option<u8>,

    /// Run both parts (overrides --part)
    #[arg(short, long)]
    pub all: bool,
}

pub struct RunConfig {
    pub all: bool,
    pub part: u8,
    pub use_sample: bool,
}

impl AocRunner {
    const SAMPLE_STR: [&'static str; 2] = ["real", "samp"];

    pub fn parse_and_run(&self) {
        let cli = Cli::parse();

        match cli.command {
            Commands::Main => self.run_full_year(),
            Commands::Day(args) => {
                let day = args.day as usize;

                if !(1..=25).contains(&day) {
                    eprintln!("Error: Day must be between 1 and 25, got: {}", day);
                    std::process::exit(1);
                }

                if day > self.solutions.len() {
                    eprintln!("No solution implemented for day {:02} in year {}", day, self.year);
                    std::process::exit(1);
                }

                let config = match self.validate_day_args(&args) {
                    Ok(cfg) => cfg,
                    Err(e) => {
                        eprintln!("Error: {}", e);
                        std::process::exit(1);
                    }
                };

                if config.all {
                    self.run_day(day);
                } else {
                    self.run_single(day, config.part, config.use_sample);
                }
            }
        }
    }

    fn validate_day_args(&self, args: &DayArgs) -> Result<RunConfig, String> {
        if args.all {
            // --all is given, part is ignored
            Ok(RunConfig {
                all: true,
                part: 1, // dummy value, not used
                use_sample: args.sample,
            })
        } else if let Some(part) = args.part {
            // --part is specified
            Ok(RunConfig {
                all: false,
                part,
                use_sample: args.sample,
            })
        } else {
            Err("--part <1|2> is required unless --all is given".to_string())
        }
    }

    pub fn run(&self, env_arg: Option<String>, full_day: bool, version: u8, use_sample: bool) {
        let arg = env_arg.expect("Pass day or 'main' as argument!");

        match arg.as_str() {
            "main" => self.run_full_year(),
            _ => {
                let day = arg
                    .strip_prefix("day")
                    .expect("Argument should start with 'day'")
                    .parse::<usize>()
                    .expect("Argument should have format 'dayXX' with XX being a valid number!");

                match full_day {
                    true => self.run_day(day),
                    false => self.run_single(day, version, use_sample),
                }
            }
        }
    }

    pub fn verify_solution(&self, day: usize, version: u8, use_sample: bool) -> Result<(), String> {
        let solution = &self.solutions[day - 1];
        assert!(version == 1 || version == 2);
        let expected = &solution.results()[(version - 1) as usize * 2 + (!use_sample as usize)];
        let (result, _) = self.get_result(day, version, use_sample);

        if result != *expected {
            Err(format!("Got {:?}, expected {:?}", result, expected))
        } else {
            Ok(())
        }
    }

    fn run_full_year(&self) {
        let mut total_elapsed = Duration::ZERO;
        for day in 0..self.solutions.len() {
            println!("Day {0:02}:", day + 1);
            let mut day_elapsed = Duration::ZERO;
            for version in [1, 2] {
                for use_sample in [true, false] {
                    let (result, elapsed) = self.get_result(day + 1, version, use_sample);
                    day_elapsed += elapsed;
                    println!("  V{version} {}:  {result}", Self::SAMPLE_STR[use_sample as usize]);
                }
            }
            total_elapsed += day_elapsed;
            println!("  > Runtime:  {day_elapsed:?}");
            println!()
        }

        println!("\n\nTotal Runtime: {total_elapsed:?}");
    }

    fn run_day(&self, day: usize) {
        let mut total_time = Duration::ZERO;
        for version in [1, 2] {
            for use_sample in [true, false] {
                let (result, elapsed) = self.get_result(day, version, use_sample);
                total_time += elapsed;
                println!(
                    "V{version} {} in {elapsed: >10.4?}:    {result}",
                    Self::SAMPLE_STR[use_sample as usize]
                );
            }
        }
        println!("\nTotal Runtime: {total_time:?}");
    }

    fn run_single(&self, day: usize, version: u8, use_sample: bool) {
        if self.solutions.len() < day {
            println!("No solution implemented for day {day:02} in year {0}", self.year);
            return;
        }

        let (result, elapsed) = self.get_result(day, version, use_sample);
        println!(
            "Day {day:02} / Version {version} / Data '{}' => {:?}\n{}",
            Self::SAMPLE_STR[use_sample as usize],
            elapsed,
            result
        );
    }

    fn get_input(&self, day: u8, version: u8, use_sample: bool) -> Option<ProblemInput> {
        let base_filename = if use_sample { "sample" } else { "input" };
        let mut fullname = format!("{0}\\inputs\\{base_filename}{day:02}.txt", self.year);
        if !Path::new(&fullname).exists() {
            fullname = fullname.replace(".txt", format!("_{version}.txt").as_str());
        }

        ProblemInput::read(&fullname)
    }

    fn get_result(&self, day: usize, version: u8, use_sample: bool) -> (ProblemResult, Duration) {
        let s = &self.solutions[day - 1];
        match self.get_input(day as u8, version, use_sample) {
            Some(input) => s.solve(input, version, use_sample),
            None => (ProblemResult::NoInput, Duration::ZERO),
        }
    }
}

File: lib/zig/runner.zig
const std = @import("std");
const solution = @import("solution.zig");
const types = @import("types.zig");

pub const RunConfig = struct {
    all: bool,
    part: u8,
    use_sample: bool,
};

pub const AocRunner = struct {
    year: u16,
    solutions: [25]?solution.Solution,
    arena_enabled: bool,
    _allocator: std.mem.Allocator = undefined,

    pub fn init(year: u16, impl_solutions: []const ?solution.Solution, use_arena: bool) AocRunner {
        var solutions: [25]?solution.Solution = .{null} ** 25;
        std.mem.copyForwards(?solution.Solution, &solutions, impl_solutions);
        return .{
            .year = year,
            .solutions = solutions,
            .arena_enabled = use_arena,
        };
    }

    pub fn parseAndRun(self: *AocRunner) !void {
        var gpa = std.heap.GeneralPurposeAllocator(.{}){};
        self._allocator = gpa.allocator();
        defer {
            const deinit_status = gpa.deinit();
            if (deinit_status == .leak) std.debug.print("LEAK", .{});
        }

        const args = try std.process.argsAlloc(self._allocator);
        defer std.process.argsFree(self._allocator, args);

        if (args.len < 2) {
            std.debug.print("Usage: {s} [main | day N --part <1|2> [--sample] [--all]]\n", .{args[0]});
            return;
        }

        const cmd = args[1];

        if (std.mem.eql(u8, cmd, "main")) {
            try self.runFullYear();
            return;
        }

        if (std.mem.eql(u8, cmd, "day")) {
            if (args.len < 3) {
                std.debug.print("Error: Day number required after 'day' command\n", .{});
                return;
            }

            const day = std.fmt.parseInt(u8, args[2], 10) catch {
                std.debug.print("Error: Invalid day number: {s}\n", .{args[2]});
                return;
            };

            if (day < 1 or day > 25) {
                std.debug.print("Error: Day must be between 1 and 25, got: {d}\n", .{day});
                return;
            }

            const config = try self.parseConfig(args[3..]);

            if (config.all) {
                try self.runDay(day);
            } else {
                try self.runSingle(day, config.part, config.use_sample);
            }
            return;
        }

        std.debug.print("Error: Unknown command: {s}. Use 'main' or 'day N'\n", .{cmd});
    }

    fn parseConfig(_: *AocRunner, args: [][]const u8) !RunConfig {
        var config = RunConfig{
            .all = false,
            .part = 0,
            .use_sample = false,
        };

        var i: usize = 0;
        while (i < args.len) : (i += 1) {
            const arg = args[i];

            if (std.mem.eql(u8, arg, "--all") or std.mem.eql(u8, arg, "-a")) {
                config.all = true;
            } else if (std.mem.eql(u8, arg, "--sample") or std.mem.eql(u8, arg, "-s")) {
                config.use_sample = true;
            } else if (std.mem.eql(u8, arg, "--part") or std.mem.eql(u8, arg, "-p")) {
                if (i + 1 >= args.len) {
                    std.debug.print("Error: --part requires a value\n", .{});
                    return error.InvalidArgs;
                }
                i += 1;
                config.part = std.fmt.parseInt(u8, args[i], 10) catch {
                    std.debug.print("Error: Invalid part number: {s}\n", .{args[i]});
                    return error.InvalidArgs;
                };
                if (config.part != 1 and config.part != 2) {
                    std.debug.print("Error: --part must be 1 or 2, got: {d}\n", .{config.part});
                    return error.InvalidArgs;
                }
            }
        }

        if (!config.all and config.part == 0) {
            std.debug.print("Error: --part <1|2> is required unless --all is given\n", .{});
            return error.InvalidArgs;
        }

        return config;
    }

    pub fn run(self: *AocRunner, full_day: bool, part: u8, use_sample: bool) !void {
        var gpa = std.heap.GeneralPurposeAllocator(.{}){};
        self._allocator = gpa.allocator();
        defer {
            const deinit_status = gpa.deinit();
            if (deinit_status == .leak) std.debug.print("LEAK", .{});
        }

        const args = try std.process.argsAlloc(self._allocator);
        defer std.process.argsFree(self._allocator, args);
        if (args.len != 2) {
            std.debug.print("Usage: {s} <day>\n", .{args[0]});
            return;
        }
        const cmd_arg = args[1];

        if (std.mem.eql(u8, cmd_arg, "main")) {
            try self.runFullYear();
            return;
        }

        if (std.fmt.parseInt(u8, cmd_arg[3..], 10)) |day| {
            if (full_day) {
                try self.runDay(day);
            } else {
                try self.runSingle(day, part, use_sample);
            }
        } else |_| {
            std.debug.print("Argument should have format dayXX: {s}\n", .{cmd_arg});
        }
    }

    pub fn verifySolutions(self: *const AocRunner) bool {
        std.debug.print("\n----------\n", .{});
        var success = true;
        for (self.solutions, 1..) |sol, day| {
            inline for ([_]u8{ 1, 2 }) |part| {
                inline for ([_]bool{ true, false }) |use_sample| {
                    const result_union = self.getResult(@as(u8, @intCast(day)), part, use_sample);
                    if (result_union) |timed_result| {
                        defer timed_result.deinit(self._allocator);
                        std.debug.print("Testing D{0d:0>2} P{1d} '{2s}': ", .{ day, part, if (use_sample) "s" else "r" });
                        const index = comptime (part - 1) * 2 + (if (use_sample) 0 else 1);
                        const expected = sol.?.results()[index];
                        if (std.testing.expect(timed_result.result.eql(expected))) |_| {
                            std.debug.print("PASSED\n", .{});
                        } else |_| {
                            std.debug.print("FAILED: Got {0s}, expected {1s}\n", .{ timed_result.result, expected });
                            success = false;
                        }
                    } else |err| switch (err) {
                        error.MissingSolution => {},
                        else => {
                            std.debug.print("Testing D{0d:0>2} P{1d} '{2s}': ", .{ day, part, if (use_sample) "s" else "r" });
                            std.debug.print("ERROR RESULT ({!})\n", .{result_union});
                            success = false;
                        },
                    }
                }
            }
        }
        std.debug.print("----------\n", .{});
        return success;
    }

    fn runFullYear(self: *const AocRunner) !void {
        var year_elapsed: f64 = 0.0;
        for (self.solutions, 1..) |_, day| {
            std.debug.print("Day {0d:0>2}\n", .{day});
            var day_elapsed: f64 = 0.0;
            for ([_]u8{ 1, 2 }) |part| {
                for ([_]bool{ true, false }) |use_sample| {
                    const timed_result = self.getResult(@as(u8, @intCast(day)), part, use_sample) catch |err| {
                        std.debug.print("  P{0d} {1s}:  {2s}\n", .{
                            part,
                            if (use_sample) "samp" else "real",
                            types.getErrorDesc(err),
                        });
                        continue;
                    };
                    defer timed_result.deinit(self._allocator);
                    day_elapsed += timed_result.duration;
                    std.debug.print("  P{0d} {1s}:  {2s}\n", .{
                        part,
                        if (use_sample) "samp" else "real",
                        timed_result.result,
                    });
                }
            }
            year_elapsed += day_elapsed;
            std.debug.print("  > Runtime: {0d}\n\n", .{day_elapsed});
        }
        std.debug.print("\n\nTotal Runtime: {0d}\n", .{year_elapsed});
    }

    fn runDay(self: *const AocRunner, day: u8) !void {
        var day_elapsed: f64 = 0.0;
        var expected_match = true;
        for ([_]u8{ 1, 2 }) |part| {
            for ([_]bool{ true, false }) |use_sample| {
                const timed_result = try self.getResult(day, part, use_sample);
                defer timed_result.deinit(self._allocator);
                day_elapsed += timed_result.duration;

                const index: usize = (part - 1) * 2 + (if (use_sample) @as(usize, 0) else @as(usize, 1));
                const expected = self.solutions[day - 1].?.results()[index];
                if (!timed_result.result.eql(expected)) {
                    expected_match = false;
                }

                std.debug.print("P{0d} {1s} in {2d:.4}s:    {3s}\n", .{
                    part,
                    if (use_sample) "samp" else "real",
                    timed_result.duration,
                    timed_result.result,
                });
            }
        }
        std.debug.print("\nTotal Runtime: {0d}s\n", .{day_elapsed});
        std.debug.print("Note: Results {s}match expected", .{if (expected_match) "" else "don't "});
    }

    fn runSingle(self: *const AocRunner, day: u8, part: u8, use_sample: bool) !void {
        const timed_result = try self.getResult(day, part, use_sample);
        defer timed_result.deinit(self._allocator);
        std.debug.print("Day {0d:0>2} / part {1d} / Data '{2s}' => {3d}s\n{4s}\n", .{
            day,
            part,
            if (use_sample) "samp" else "real",
            timed_result.duration,
            timed_result.result,
        });
    }

    fn getResult(self: *const AocRunner, day: u8, part: u8, use_sample: bool) types.SolvingError!types.TimedResult {
        if (day > 25 or day == 0) {
            std.debug.print("AoC only has 25 days: {d}\n", .{day});
            return types.SolvingError.InvalidDay;
        }
        const s = self.solutions[day - 1];
        if (s == null) {
            return types.SolvingError.MissingSolution;
        }

        var input = self.getInput(day, part, use_sample);
        if (input == null) {
            return .{ .result = types.Result.NoInput, .duration = 0 };
        }

        defer input.?.deinit();

        // Use arena allocator in solutions for easier coding
        if (self.arena_enabled) {
            var arena = std.heap.ArenaAllocator.init(self._allocator);
            const solution_allocator = arena.allocator();
            defer arena.deinit();

            // In case result contains heap allocated memory, clone with gpa before freeing
            var original_result = try s.?.solve(&input.?, part, use_sample, solution_allocator);
            return original_result.clone(self._allocator) catch types.SolvingError.SolvingFailed;
        } else {
            return s.?.solve(&input.?, part, use_sample, self._allocator);
        }
    }

    fn getInput(self: *const AocRunner, day: u8, part: u8, use_sample: bool) ?types.ProblemInput {
        const base_filename = if (use_sample) "sample" else "input";
        const filename = std.fmt.allocPrint(self._allocator, "{0d}\\inputs\\{1s}{2d:0>2}.txt", .{ self.year, base_filename, day }) catch return null;
        defer self._allocator.free(filename);

        const input = types.ProblemInput.read(self._allocator, filename) catch |err| switch (err) {
            std.fs.File.OpenError.FileNotFound => {
                const filename_vers = std.fmt.allocPrint(self._allocator, "{0d}\\inputs\\{1s}{2d:0>2}_{3d}.txt", .{ self.year, base_filename, day, part }) catch unreachable;
                defer self._allocator.free(filename_vers);
                return types.ProblemInput.read(self._allocator, filename_vers) catch return null;
            },
            else => return null,
        };

        return input;
    }
};
